lib/api/CDataFrameAnalysisInstrumentation.cc

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <api/CDataFrameAnalysisInstrumentation.h>
+
+namespace ml {
+namespace api {
+
+namespace {
+const std::string STEP_TAG{"step"};
+const std::string PROGRESS_TAG{"progress"};
+const std::string PEAK_MEMORY_USAGE_TAG{"peak_memory_usage"};
+
+const std::size_t MAXIMUM_FRACTIONAL_PROGRESS{std::size_t{1}
+                                              << ((sizeof(std::size_t) - 2) * 8)};
+}
+
+void CDataFrameAnalysisInstrumentation::updateMemoryUsage(std::int64_t delta) {
+    std::int64_t memory{m_Memory.fetch_add(delta)};
+    if (memory >= 0) {
+        core::CProgramCounters::counter(this->memoryCounterType()).max(memory);
+    } else {
+        // Something has gone wrong with memory estimation. Trap this case
+        // to avoid underflowing the peak memory usage statistic.
+        LOG_WARN(<< "Memory estimate " << memory << " is negative!");
+    }
+}
+
+void CDataFrameAnalysisInstrumentation::updateProgress(double fractionalProgress) {
+    m_FractionalProgress.fetch_add(static_cast<std::size_t>(std::max(
+        static_cast<double>(MAXIMUM_FRACTIONAL_PROGRESS) * fractionalProgress + 0.5, 1.0)));
+}
+
+void CDataFrameAnalysisInstrumentation::setToFinished() {
+    m_Finished.store(true);
+    m_FractionalProgress.store(MAXIMUM_FRACTIONAL_PROGRESS);
+}
+
+bool CDataFrameAnalysisInstrumentation::finished() const {
+    return m_Finished.load();
+}
+
+double CDataFrameAnalysisInstrumentation::progress() const {
+    return this->finished()
+               ? 1.0
+               : static_cast<double>(std::min(m_FractionalProgress.load(),
+                                              MAXIMUM_FRACTIONAL_PROGRESS - 1)) /
+                     static_cast<double>(MAXIMUM_FRACTIONAL_PROGRESS);
+}
+
+CDataFrameAnalysisInstrumentation::CDataFrameAnalysisInstrumentation()
+    : m_Finished{false}, m_FractionalProgress{0}, m_Memory{0}, m_Writer{nullptr} {
+}
+
+void CDataFrameAnalysisInstrumentation::resetProgress() {
+    m_FractionalProgress.store(0.0);
+    m_Finished.store(false);
+}
+
+void CDataFrameAnalysisInstrumentation::writer(core::CRapidJsonConcurrentLineWriter* writer) {
+    m_Writer = writer;
+}
+
+void CDataFrameAnalysisInstrumentation::nextStep(std::uint32_t step) {
+    m_StateQueue.tryPush(SInternalState(*this));
+    if (m_Writer != nullptr) {
+        while (m_StateQueue.size() > 0) {
+            this->writeState(step, m_StateQueue.pop());
+        }
+    }
```

Comment:
Ok, I'm now wondering about the use of the queue here at all. These operations all happen on the same thread. Currently, there is no protection that the writer isn't written to by multiple things at once. Although in practice `nextStep` is called by single threaded code in the actual runners we also write progress from a different thread in a loop in `CDataFrameAnalyzer`. However, this uses a concurrent line writer so each JSON object is written to a single line and it manages writes from different threads. So I think we can just remove the queue altogether. Also, [this](https://github.com/elastic/ml-cpp/blob/master/lib/api/CDataFrameAnalyzer.cc#L138) todo can then be removed since it does need to support concurrent writes.

Reply:
I changed the code as discussed